stdlib/dynarray.mli

```diff
@@ -0,0 +1,619 @@
+    a change.
+
+    {!to_seq} produces an on-demand sequence of values, and is expected
+    to be called with effects happening in-between. Its specification
+    tolerates changes of length. (See below.)
+*)
+
+val of_array : 'a array -> 'a t
+(** [of_array arr] returns a dynamic array corresponding to the fixed-sized array [a].
+    Operates in [O(n)] time by making a copy. *)
+
+val to_array : 'a t -> 'a array
+(** [to_array a] returns a fixed-sized array corresponding to the dynamic array [a].
+    This always allocate a new array and copies elements into it. *)
+
+val of_list : 'a list -> 'a t
+(** [of_list l] is the array containing the elements of [l] in
+    the same order. *)
+
+val to_list : 'a t -> 'a list
+(** [to_list a] is a list with the elements contained in the array [a]. *)
+
+val of_seq : 'a Seq.t -> 'a t
+(** [of_seq seq] is an array containing the same elements as [seq].
+
+    It traverses [seq] once and will terminate only if [seq] is finite. *)
+
+val to_seq : 'a t -> 'a Seq.t
+(** [to_seq a] is the sequence of elements
+    [get a 0], [get a 1]... [get a (length a - 1)].
+
+    Because sequences are computed on-demand, we have to assume that
+    the array may be modified in the meantime. We give a precise
+    specification for this case below.
+
+    Demanding the [i]-th element of the resulting sequence (which can
+    happen zero, one or several times) will access the [i]-th element
+    of [a] at the time of the demand. The sequence stops if [a] has
+    less than [i] elements at this point.
+*)
+
+val to_seq_rev : 'a t -> 'a Seq.t
+(** [to_seq_rev a] is the sequence of elements
+    [get a (l - 1)], [get a (l - 2)]... [get a 0],
+    where [l] is [length a] at the time [to_seq_rev] is invoked.
+
+    Elements that have been removed from the array by the time they
+    are demanded in the sequence are skipped.
+*)
+
+
+(** {1:advanced Advanced topics for performance} *)
+
+(** {2:capacity Backing array, capacity}
+
+    Internally, a dynamic array uses a {b backing array} (a fixed-size
+    array as provided by the {!Array} module) whose length is greater
+    or equal to the length of the dynamic array. We call {b capacity}
+    the length of the backing array.
+
+    The capacity of a dynamic array is relevant in advanced scenarios,
+    when reasoning about the performance of dynamic array programs:
+    {ul
+    {- The memory usage of a dynamic array is proportional to its capacity,
+       rather than its length.}
+    {- When then is no empty space left at the end of the backing array.
+       adding elements requires allocating a new, larger backing array.}}
+
+    The implementation uses a standard exponential reallocation
+    strategy which guarantees amortized constant-time operation: the
+    total capacity of all backing arrays allocated over the lifetime
+    of a dynamic array is at worst proportional to the total number of
+    elements added.
```

Comment:
As far as I can tell, this part is correct:

> the total capacity of all backing arrays allocated over the lifetime
    of a dynamic array is at worst proportional to the total number of
    elements added.

Counting the number of elements added is clearer, I think, than counting the maximal past number of elements, and equivalent in the worst case (where we only add elements). It is more precise than your proposal, because it bounds the total allocated size, not the current size. (For example, an implementation that would suffer from repeated upsizing/resizing when repeating pushes and pop at a specific index would satisfy your proprety but violate this one.)

I think that you are unhappy with a claim of constant-time followed by a space-related description. But here we are talking  about the time spent in reallocation/resizing, which is proportional to their total allocated words.

Reply:
Amortized time complexity is always about the average running time of a sequence of operations. That includes the complete execution time of the operations, not just the time spent allocating. An implementation with an exponential time complexity that never allocates would satisfy your definition, but would be bad. What you're describing is more "amortized constant-allocating operations" (where each operation in the sequence allocates a constant amount on average). It's a nice property to have, but not what most users are interested in, and it's a weaker claim than constant-time (you can only allocate a constant amount in constant time, so constant-time implies constant-allocating).

Reply:
I am frustrated by the way this conversation is going.

I think that it is time to take a step back and explain that my *intention* after the `:` was *not* to give a formal definition of what amortized constant time means (either the reader knows, and they don't need one, or they don't, and now is not the right time to patronize them about it), but a novice-level explanation in less technical words of the performance model they should have in mind when thinking about adding elements to a dynarray. I understand that my intention has not been reflected properly by my text. I will rephrase accordingly, thanks!

The fact that constant-time means constant-allocated-bytes does *not* imply that explaining things in terms of time is going to be easier for users to understand, in fact it makes it strictly more confusing. Also, I would be careful before making precise claims on "the complete execution time of operations" as you suggest, because we are in a GC-ed language where any allocation may have a latency that is *not* bounded by some function of the number of add/delete operations that happened on the dynarray so far.

Reply:
I'm sorry for causing frustration, that was not my intention. The `:` really made it look like you were trying to explain the meaning of amortized constant time. If you just use a `.` instead, it's already clear that you're stating two separate facts.